### DISCUSSION TITLE
Remove Non-Inclusive Terms

### DISCUSSION THREAD

**Discussion by tinder-owenthomas** [CocoaLumberjack/CocoaLumberjack#1362] [General]
### Issue Info

CocoaLumberjack currently has several references to non-inclusive terms.

### Issue Description and Steps

Can we please get an updated version of the SDK which removes all non-inclusive language? We are focusing on removing such language from our company which extends to all third-party code we choose to use.

Specifically terms such as [whitelist](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=whitelist) (also [here](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=addToWhitelist), [here](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=removeFromWhitelist), and [here](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=isOnWhitelist)), [blacklist](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=blacklist) (also [here](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=addtoblacklist), and [here](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=removefromblacklist)), [master](https://github.com/CocoaLumberjack/CocoaLumberjack/search?q=master), slave, etc. Would love this to cover actual code of course, but also, comments, branch names, repo descriptions on dependency management central repositories, licensing agreements, etc.

**Comment by ffried** [comment]
Thanks for bringing this to our attention!
IMHO this needs a team discussion here (pinging @CocoaLumberjack/collaborators).
<hr>
My two cents:

I think CocoaLumberjack doesn't have too many references. The blacklist/whitelist ones should be rather straightforward to deprecate and replace with blocklist/allowlist (<- my suggestion for the new names).

As for `master`: My personal opinion here is that in the context of `git` renaming `master` is nonsense. While I don't agree with everything, [this blog post](https://felipec.wordpress.com/2020/11/13/git-master/) lists some IMHO important points as to why. But from a more objective perspective: Renaming `master` is non-trivial and breaks many integrations. It's not impossible, but it's needs careful consideration. My personal vote goes against it.

Apart from `git` / branches:
There's one `#pragma mark` containing "master". That one's trivial and might not fit the points described in the blog post.
However, there are also many references to `sqlite_master`, which is not something we can change.

**Comment by lolgear** [comment]
Please, stop this madness about naming conventions.

Instead of renaming everything, you could be more "useful" and find bugs in this repository. Or add tests. Or add support for Windows (WHAT?).

----

My two cents.

If you are a developer who cares about people experience with your application, you will follow the conventions of the world. You don't care about "not right name" in fundamental tools that you use.

I write these words in English, but I would like to use any language. Is it offensive to me? Does it diminish my rights and my country? Will people tell me if I make a mistake? (But they shouldn't, because it is offensive to me. Knowledge is offensive.)

Do you know about another thing - the famous equation that rules our world?
Look at it. [Black–Scholes equation](https://en.wikipedia.org/wiki/Black–Scholes_equation).

Oh no! It contains B. word but it is a surname of famous economist.
Should we banned his surname and rename equation?

**Comment by sushichop** [comment]
Here's my two cents.

I think it's basically okay to rename words that can be renamed easily. 
However, I think it is actually difficult to rename `master` because of its large impact.

I straightforwardly agree that we should name the word carefully when we create a new repository. 
On the other hand, IMHO, it is important to consider the impact of changes on existing repositories...

**Comment by tinder-owenthomas** [comment]
I understand where you are coming from @lolgear. As a very privileged white, abled, American, male, etc. I too used to feel that efforts like this were not important - because they didn't impact me so I didn't see or understand them. But over the past several years I've come to realize just how important they are to others, and therefore to our society more broadly, myself included. I have met frequently with people for whom the use of non-inclusive terminology is hurtful, demeaning, and marginalizing, and came to understand how truly important this is. I'd urge you to go out and talk to some marginalized groups at your workplace, friend group, the broader iOS engineering community, etc. and hear directly from them how a culture of exclusion has impacted them in ways we may never fully understand. Just because it doesn't impact you and me directly to the same extent, doesn't mean it is unimportant.

Separately, this is an important initiative for my company as we believe in fully embracing diversity, equity, and inclusion. Terms like master, slave, whitelist, blacklist, etc. in use in technology are in fact directly derived from slavery and/or other historical and societal forms of oppression and discrimination - even though it may not seem like it, and even though these terms are generally not used consciously in that way by engineers today. This is easy to confuse with other uses of the same word as in your example of the Black-Scholes model - which is perfectly fine as it is referring to a person's name and not a subordinate status to the "white" alternative. However, it is hard to deny that using a word like "blacklist" to mean exclusionary, bad, rejected, etc. while a "whitelist" on the other hand means accepted, allowed, or preferred is clearly derived from race relations and not a random choice of words that was brand new in the tech sector. To be very clear I am not saying you cannot use the words for their proper meaning: feel free to paint your walls "white", drive a "black" car, or watch "black and white" films. Those are fine but many other terms we use frequently in technology are highly problematic. Unfortunately, the continued use of these terms - even unintentionally - by respected companies and individuals is highly offensive to groups of people who are already marginalized in so many other ways and continues and reinforces the culture of oppression that we have today. You and I may not see it because we likely benefit from this culture, but it comes at a great cost to others.

On a more personal level, these terms are triggering for me now that I have learned more about their impact on individuals and society in general - not to mention my direct co-workers whom I greatly respect, and the members who use our company's technology on a daily basis. Refusing to make these types of simple changes now that we know better is a direct slap in the face to all these people, which I find unacceptable.

For all the above reasons we are actively working to correct past mistakes by removing non-inclusive terminology, and we are far from alone in this effort. Many other large reputable software engineering groups are doing the same thing in an attempt to level the playing field for all people: [Apple](https://appleinsider.com/articles/20/07/17/apple-to-remove-replace-non-inclusive-language-in-code-base), [Github](https://www.zdnet.com/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references/), [IBM & Microsoft](https://www.zdnet.com/article/ibm-microsoft-staff-rally-to-remove-racially-insensitive-language-from-products/), [Twitter & JPMorgan](https://www.cnn.com/2020/07/03/tech/twitter-jpmorgan-slave-master-coding/index.html), [The Linux Kernel Team](https://itsfoss.com/linux-kernel-inclusive-code), [MySQL](https://www.zdnet.com/article/mysql-drops-master-slave-and-blacklist-whitelist-terminology/), and hundreds of others.

I hope through discussions like this you can come to appreciate the importance of this effort, even if only for others, and help be part of the change that is essential to move our society - and software engineering much more specifically - in the right direction.

**Comment by bpoplauschi** [comment]
@tinder-owenthomas I really respect you taking the time to explain this situation, especially since Github brings together people with different backgrounds, some of which (like me) are not really exposed to the effects of racism.

@CocoaLumberjack/collaborators I think this is important. If by a few small changes, we can make at least a few people not being reminded about racism, I say let's do it. I say this is an opportunity to be more open-minded.
I'm not that scared of the renaming of `master` to `main` effects. For starters, we can keep both, but make our tools work with `main`.

**Comment by sushichop** [comment]
@tinder-owenthomas @bpoplauschi 
Thanks for explaining the important things in detail. I basically agree with you.
Then I think [the following page](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/diff/?id=49decddd39e5f6132ccd7d9fdc3d7c470b0061bb) will be helpful to change the words.

```
Recommended replacements for 'master / slave' are:
    '{primary,main} / {secondary,replica,subordinate}'
    '{initiator,requester} / {target,responder}'
    '{controller,host} / {device,worker,proxy}'
    'leader / follower'
    'director / performer'

Recommended replacements for 'blacklist/whitelist' are:
    'denylist / allowlist'
    'blocklist / passlist'
```

For starters, I will create a PR to change `blacklist/whitelist` to `denylist/allowlist` because I think there is no reason why these words cannot be changed 🙂

**Comment by sushichop** [comment]
I have created a PR that changes `blacklist/whitelist` to `denylist/allowlist`.

Then, as a next step, what should we do with `sqlite_master`? 
About branch name, IMO, we can change it to `main` or use both `main` and `master` branches temporarily...

**Comment by lolgear** [comment]
@sushichop 
If I don't like something at GitHub, I will open PR  and ask people to review it.
And, possibly, I will create an issue. ( any order, but it is better to create PR first ). ( Discussion in PR is possible, so, opening issue may be extra effort ).
This approach is effective and you will get respect of all contributors.

I don't appreciate idea of whining.

**Comment by sushichop** [comment]
@lolgear 
Thanks your idea🙂

**Comment by ffried** [comment]
The code part of this (#1218) just went live in 3.7.1 🙂

**Comment by HermanKayy** [comment]
@sushichop @ffried Thanks for doing this! It looks like in DDContextFilterLogFormatter+Deprecated.h and DDContextFilterLogFormatter+Deprecated.m still have a few places leftover that's referencing blacklist and whitelist. I know those files are deprecated but any chance to rename them still?

**Comment by ffried** [comment]
> @sushichop @ffried Thanks for doing this! It looks like in DDContextFilterLogFormatter+Deprecated.h and DDContextFilterLogFormatter+Deprecated.m still have a few places leftover that's referencing blacklist and whitelist. I know those files are deprecated but any chance to rename them still?

@HermanKayy These were newly deprecated in #1218. We can't completely remove/rename those APIs without causing some major problems with existing users. We need to treat this change the same as all other API deprecation. We can eventually remove it, but we need to keep backwards compatibility for now.
